File: src/utils/template.cpp
#include<bits/stdc++.h>
using namespace std;

// Definition for singly-linked list
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

// Definition for binary tree
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

namespace IO {
    /**
     * Input functions for various data types
     * Each function handles reading from stdin for a specific type
     */
    namespace Input {
        // Basic types
        void read(bool& x) { cin >> x; }
        void read(char& x) { cin >> x; }
        void read(int& x) { cin >> x; }
        void read(unsigned int& x) { cin >> x; }
        void read(long int& x) { cin >> x; }
        void read(unsigned long& x) { cin >> x; }
        void read(long long int& x) { cin >> x; }
        void read(unsigned long long& x) { cin >> x; }
        void read(float& x) { cin >> x; }
        void read(double& x) { cin >> x; }
        void read(long double& x) { cin >> x; }
        
        // String with whitespace handling
        void read(string& x) { 
            cin >> ws; // Consume leading whitespace
            getline(cin, x); 
        }
        
        // Linked list
        void read(ListNode*& head) {
            int n; 
            cin >> n;
            if (n == 0) { 
                head = nullptr; 
                return; 
            }
            vector<int> values(n);
            for (int i = 0; i < n; i++) { 
                cin >> values[i]; 
            }
            head = new ListNode(values[0]);
            ListNode* current = head;
            for (int i = 1; i < n; i++) {
                current->next = new ListNode(values[i]);
                current = current->next;
            }
        }
        
        // Binary tree (level-order traversal format)
        void read(TreeNode*& root) {
            int n; 
            cin >> n;
            if (n == 0) { 
                root = nullptr; 
                return;
            }
            vector<string> values(n);
            for (int i = 0; i < n; i++) {
                cin >> values[i]; 
            }
            if (values[0] == "null") {
                root = nullptr; 
                return;
            }
            root = new TreeNode(stoi(values[0]));
            queue<TreeNode*> q;
            q.push(root);
            int i = 1;
            while (!q.empty() && i < n) {
                TreeNode* node = q.front();
                q.pop();
                if (i < n) {
                    if (values[i] != "null") {
                        node->left = new TreeNode(stoi(values[i]));
                        q.push(node->left);
                    }
                    i++;
                }
                if (i < n) {
                    if (values[i] != "null") {
                        node->right = new TreeNode(stoi(values[i]));
                        q.push(node->right);
                    }
                    i++;
                }
            }
        }
        // Forward declarations for template specializations
        template <typename T> void read(vector<T>& x);
        template <typename F, typename S> void read(pair<F, S>& x);
        template <typename K, typename V> void read(map<K, V>& x);
        template <typename K, typename V> void read(unordered_map<K, V>& x);
        template <typename T> void read(set<T>& x);
        template <typename T> void read(unordered_set<T>& x);
        
        // Vector
        template <typename T>
        void read(vector<T>& x) {
            int n;
            cin >> n;
            x.resize(n);
            for (int i = 0; i < n; i++) {
                read(x[i]);
            }
        }
        
        // Pair
        template <typename F, typename S>
        void read(pair<F, S>& x) {
            read(x.first);
            read(x.second);
        }
        
        // Map
        template <typename K, typename V>
        void read(map<K, V>& x) {
            int n;
            cin >> n;
            for (int i = 0; i < n; i++) {
                K key;
                V value;
                read(key);
                read(value);
                x[key] = value;
            }
        }
        
        // Unordered Map
        template <typename K, typename V>
        void read(unordered_map<K, V>& x) {
            int n;
            cin >> n;
            for (int i = 0; i < n; i++) {
                K key;
                V value;
                read(key);
                read(value);
                x[key] = value;
            }
        }
        
        // Set
        template <typename T>
        void read(set<T>& x) {
            int n;
            cin >> n;
            for (int i = 0; i < n; i++) {
                T value;
                read(value);
                x.insert(value);
            }
        }
        
        // Unordered Set
        template <typename T>
        void read(unordered_set<T>& x) {
            int n;
            cin >> n;
            for (int i = 0; i < n; i++) {
                T value;
                read(value);
                x.insert(value);
            }
        }
    } // namespace Input
    
    /**
     * Output functions for various data types
     * Each function handles printing to stdout for a specific type
     */
    namespace Output {
        // Basic types
        void write(bool x) { cout << (x ? "True" : "False"); }
        void write(char x) { cout << x; }
        void write(int x) { cout << x; }
        void write(unsigned int x) { cout << x; }
        void write(long int x) { cout << x; }
        void write(unsigned long x) { cout << x; }
        void write(long long int x) { cout << x; }
        void write(unsigned long long x) { cout << x; }
        void write(float x) { cout << x; }
        void write(double x) { cout << x; }
        void write(long double x) { cout << x; }
        void write(string x) { cout << x; }
        void write(const char* x) { cout << x; }
        
        // Linked list
        void write(ListNode* head) {
            while (head) {
                cout << head->val;
                if (head->next) cout << " -> ";
                head = head->next;
            }
        }
        
        // Binary tree (level-order output)
        void write(TreeNode* root) {
            if (!root) {
                cout << "[]";
                return;
            }
            cout << "[";
            queue<TreeNode*> q;
            q.push(root);
            bool first = true;
            while (!q.empty()) {
                int levelSize = q.size();
                bool hasNextLevel = false;
                for (int i = 0; i < levelSize; i++) {
                    TreeNode* node = q.front();
                    q.pop();
                    if (!first) cout << ", ";
                    first = false;
                    if (node) {
                        cout << node->val;
                        q.push(node->left);
                        q.push(node->right);
                        if (node->left || node->right) hasNextLevel = true;
                    } else {
                        cout << "null";
                    }
                }
                if (!hasNextLevel) break;
            }
            cout << "]";
        }
        
        // Forward declarations for template specializations
        template <typename T> void write(const vector<T>& x);
        template <typename F, typename S> void write(const pair<F, S>& x);
        template <typename K, typename V> void write(const map<K, V>& x);
        template <typename K, typename V> void write(const unordered_map<K, V>& x);
        template <typename T> void write(const set<T>& x);
        template <typename T> void write(const unordered_set<T>& x);
        
        // Vector
        template <typename T>
        void write(const vector<T>& x) {
            cout << "[";
            int n = x.size();
            for (int i = 0; i < n; i++) {
                write(x[i]);
                if (i < n - 1) cout << ", ";
            }
            cout << "]";
        }
        
        // Pair
        template <typename F, typename S>
        void write(const pair<F, S>& x) {
            cout << "(";
            write(x.first);
            cout << ", ";
            write(x.second);
            cout << ")";
        }
        
        // Map
        template <typename K, typename V>
        void write(const map<K, V>& x) {
            cout << "{";
            bool first = true;
            for (const auto& [k, v] : x) {
                if (!first) cout << ", ";
                first = false;
                write(k);
                cout << ": ";
                write(v);
            }
            cout << "}";
        }
        
        // Unordered map
        template <typename K, typename V>
        void write(const unordered_map<K, V>& x) {
            cout << "{";
            bool first = true;
            for (const auto& [k, v] : x) {
                if (!first) cout << ", ";
                first = false;
                write(k);
                cout << ": ";
                write(v);
            }
            cout << "}";
        }
        
        // Set
        template <typename T>
        void write(const set<T>& x) {
            cout << "{";
            bool first = true;
            for (const auto& elem : x) {
                if (!first) cout << ", ";
                first = false;
                write(elem);
            }
            cout << "}";
        }
        
        // Unordered set
        template <typename T>
        void write(const unordered_set<T>& x) {
            cout << "{";
            bool first = true;
            for (const auto& elem : x) {
                if (!first) cout << ", ";
                first = false;
                write(elem);
            }
            cout << "}";
        }
    } // namespace Output
    
    // Convenience wrapper functions to match the original interface
    template <typename T>
    void input(T& x) {
        Input::read(x);
    }
    
    template <typename T>
    void output(T x) {
        Output::write(x);
        cout << " ";
    }
} 

using namespace IO;

{{Solution Class}}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    
    int t;
    input(t);
    while (t--) {
{{Input Statements}}
    }
    
    return 0;
}

File: src/core/languages/python/template.py
from typing import *
from collections import deque, defaultdict
import random

# Global input buffer for single word reading
class InputBuffer:
    def __init__(self):
        self.buffer = []
        self.index = 0
    
    def next_word(self):
        while self.index >= len(self.buffer):
            try:
                line = input().strip()
                if line:
                    self.buffer = line.split()
                    self.index = 0
                else:
                    continue
            except EOFError:
                return None
        
        word = self.buffer[self.index]
        self.index += 1
        return word
    
    def next_line(self):
        self.buffer = []
        self.index = 0
        try:
            return input().strip()
        except EOFError:
            return None

_input_buffer = InputBuffer()

# Definition for singly-linked list
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

# Definition for binary tree node
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class IO:
    """
    Input/Output utility class for LeetCode problems
    Handles reading from stdin and writing to stdout for various data types
    """
    
    class Input:
        """Input functions for various data types"""
        
        @staticmethod
        def read_bool() -> bool:
            word = _input_buffer.next_word()
            return word.lower() in ['true', '1', 'yes'] if word else False
        
        @staticmethod
        def read_int() -> int:
            word = _input_buffer.next_word()
            return int(word) if word else 0
        
        @staticmethod
        def read_float() -> float:
            word = _input_buffer.next_word()
            return float(word) if word else 0.0
        
        @staticmethod
        def read_string() -> str:
            return _input_buffer.next_line()
        
        @staticmethod
        def read_char() -> str:
            word = _input_buffer.next_word()
            return word[0] if word else ''
        
        @staticmethod
        def consume_newline() -> None:
            _input_buffer.next_line()
        
        # Array functions
        @staticmethod
        def read_int_array() -> List[int]:
            n = IO.Input.read_int()  
            return [IO.Input.read_int() for _ in range(n)] 
        
        @staticmethod
        def read_float_array() -> List[float]:
            n = IO.Input.read_int()  
            return [IO.Input.read_float() for _ in range(n)]  
        
        @staticmethod
        def read_string_array() -> List[str]:
            n = IO.Input.read_int() 
            return [IO.Input.read_string() for _ in range(n)]  
        
        @staticmethod
        def read_char_array() -> List[str]:
            return list(IO.Input.read_string())
        
        # 2D Array functions
        @staticmethod
        def read_int_2d_array() -> List[List[int]]:
            m = IO.Input.read_int()  
            result = []
            for _ in range(m):
                n = IO.Input.read_int()  
                row = [IO.Input.read_int() for _ in range(n)] 
                result.append(row)
            return result
        
        @staticmethod
        def read_char_2d_array() -> List[List[str]]:
            m = IO.Input.read_int() 
            result = []
            for _ in range(m):
                row_str = IO.Input.read_string()  
                result.append(list(row_str))
            return result
        
        # Set and Dictionary functions
        @staticmethod
        def read_int_set() -> Set[int]:
            n = IO.Input.read_int()
            return {IO.Input.read_int() for _ in range(n)}
        
        @staticmethod
        def read_string_set() -> Set[str]:
            n = IO.Input.read_int()
            return {_input_buffer.next_word() for _ in range(n)}
        
        @staticmethod
        def read_int_int_dict() -> Dict[int, int]:
            n = IO.Input.read_int()
            result = {}
            for _ in range(n):
                key = IO.Input.read_int()
                value = IO.Input.read_int()
                result[key] = value
            return result
        
        @staticmethod
        def read_string_int_dict() -> Dict[str, int]:
            n = IO.Input.read_int()
            result = {}
            for _ in range(n):
                key = _input_buffer.next_word()
                value = IO.Input.read_int()
                result[key] = value
            return result
        
        # Linked List
        @staticmethod
        def read_list_node() -> Optional[ListNode]:
            n = IO.Input.read_int()
            if n == 0:
                return None
            
            values = [IO.Input.read_int() for _ in range(n)]
            if not values:
                return None
                
            head = ListNode(values[0])
            current = head
            for i in range(1, len(values)):
                current.next = ListNode(values[i])
                current = current.next
            return head
        
        # Binary Tree (level-order traversal format)
        @staticmethod
        def read_tree_node() -> Optional[TreeNode]:
            n = IO.Input.read_int()
            if n == 0:
                return None
                
            values = []
            for _ in range(n):
                word = _input_buffer.next_word()
                if word and word.lower() == 'null':
                    values.append(None)
                else:
                    values.append(int(word) if word else 0)
            
            if not values or values[0] is None:
                return None
            
            root = TreeNode(values[0])
            queue = deque([root])
            i = 1
            
            while queue and i < len(values):
                node = queue.popleft()
                
                if i < len(values):
                    if values[i] is not None:
                        node.left = TreeNode(values[i])
                        queue.append(node.left)
                    i += 1
                
                if i < len(values):
                    if values[i] is not None:
                        node.right = TreeNode(values[i])
                        queue.append(node.right)
                    i += 1
            
            return root
        
        # Tuple functions
        @staticmethod
        def read_int_pair() -> Tuple[int, int]:
            first = IO.Input.read_int()
            second = IO.Input.read_int()
            return (first, second)
        
        @staticmethod
        def read_string_pair() -> Tuple[str, str]:
            first = _input_buffer.next_word()
            second = _input_buffer.next_word()
            return (first, second)
    
    class Output:
        """Output functions for various data types"""
        
        @staticmethod
        def write_bool(x: bool) -> None:
            print("true" if x else "false", end="")
        
        @staticmethod
        def write_int(x: int) -> None:
            print(x, end="")
        
        @staticmethod
        def write_float(x: float) -> None:
            print(x, end="")
        
        @staticmethod
        def write_string(x: str) -> None:
            print(f'"{x}"', end="")
        
        @staticmethod
        def write_char(x: str) -> None:
            print(x, end="")
        
        # Array functions
        @staticmethod
        def write_int_array(arr: List[int]) -> None:
            print("[", end="")
            for i, val in enumerate(arr):
                print(val, end="")
                if i < len(arr) - 1:
                    print(",", end="")
            print("]", end="")
        
        @staticmethod
        def write_float_array(arr: List[float]) -> None:
            print("[", end="")
            for i, val in enumerate(arr):
                print(val, end="")
                if i < len(arr) - 1:
                    print(",", end="")
            print("]", end="")
        
        @staticmethod
        def write_string_array(arr: List[str]) -> None:
            print("[", end="")
            for i, val in enumerate(arr):
                print(f'"{val}"', end="")
                if i < len(arr) - 1:
                    print(",", end="")
            print("]", end="")
        
        @staticmethod
        def write_char_array(arr: List[str]) -> None:
            print("[", end="")
            for i, val in enumerate(arr):
                print(f"'{val}'", end="")
                if i < len(arr) - 1:
                    print(",", end="")
            print("]", end="")
        
        # 2D Array functions
        @staticmethod
        def write_int_2d_array(arr: List[List[int]]) -> None:
            print("[", end="")
            for i, row in enumerate(arr):
                IO.Output.write_int_array(row)
                if i < len(arr) - 1:
                    print(",", end="")
            print("]", end="")
        
        @staticmethod
        def write_char_2d_array(arr: List[List[str]]) -> None:
            print("[", end="")
            for i, row in enumerate(arr):
                IO.Output.write_char_array(row)
                if i < len(arr) - 1:
                    print(",", end="")
            print("]", end="")
        
        # Set and Dictionary functions
        @staticmethod
        def write_set(s: Set) -> None:
            print("{", end="")
            items = list(s)
            for i, val in enumerate(items):
                print(val, end="")
                if i < len(items) - 1:
                    print(",", end="")
            print("}", end="")
        
        @staticmethod
        def write_dict(d: Dict) -> None:
            print("{", end="")
            items = list(d.items())
            for i, (key, val) in enumerate(items):
                print(f"{key}: {val}", end="")
                if i < len(items) - 1:
                    print(",", end="")
            print("}", end="")
        
        # Linked List
        @staticmethod
        def write_list_node(head: Optional[ListNode]) -> None:
            current = head
            while current:
                print(current.val, end="")
                if current.next:
                    print(" -> ", end="")
                current = current.next
        
        # Binary Tree (level-order output)
        @staticmethod
        def write_tree_node(root: Optional[TreeNode]) -> None:
            if not root:
                print("[]", end="")
                return
            
            print("[", end="")
            queue = deque([root])
            first = True
            
            while queue:
                level_size = len(queue)
                has_next_level = False
                
                for _ in range(level_size):
                    node = queue.popleft()
                    
                    if not first:
                        print(", ", end="")
                    first = False
                    
                    if node:
                        print(node.val, end="")
                        queue.append(node.left)
                        queue.append(node.right)
                        if node.left or node.right:
                            has_next_level = True
                    else:
                        print("null", end="")
                
                if not has_next_level:
                    break
            
            print("]", end="")
        
        # Tuple functions
        @staticmethod
        def write_tuple(t: tuple) -> None:
            print("(", end="")
            for i, val in enumerate(t):
                print(val, end="")
                if i < len(t) - 1:
                    print(", ", end="")
            print(")", end="")
    
    # Convenience wrapper functions
    @staticmethod
    def output(x) -> None:
        """
        Generic output function that determines the type and calls appropriate write function
        """
        if x is None:
            print("null", end="")
        elif isinstance(x, bool):
            IO.Output.write_bool(x)
        elif isinstance(x, int):
            IO.Output.write_int(x)
        elif isinstance(x, float):
            IO.Output.write_float(x)
        elif isinstance(x, str) and len(x) == 1:
            IO.Output.write_char(x)
        elif isinstance(x, str):
            IO.Output.write_string(x)
        elif isinstance(x, list):
            if x and isinstance(x[0], list):
                # 2D array
                if x[0] and isinstance(x[0][0], int):
                    IO.Output.write_int_2d_array(x)
                elif x[0] and isinstance(x[0][0], str):
                    IO.Output.write_char_2d_array(x)
                else:
                    print(x, end="")
            elif x and isinstance(x[0], int):
                IO.Output.write_int_array(x)
            elif x and isinstance(x[0], float):
                IO.Output.write_float_array(x)
            elif x and isinstance(x[0], str):
                if all(len(s) == 1 for s in x):
                    IO.Output.write_char_array(x)
                else:
                    IO.Output.write_string_array(x)
            else:
                print(x, end="")
        elif isinstance(x, set):
            IO.Output.write_set(x)
        elif isinstance(x, dict):
            IO.Output.write_dict(x)
        elif isinstance(x, ListNode):
            IO.Output.write_list_node(x)
        elif isinstance(x, TreeNode):
            IO.Output.write_tree_node(x)
        elif isinstance(x, tuple):
            IO.Output.write_tuple(x)
        else:
            print(x, end="")

# Convenience functions for backward compatibility
def input_int() -> int:
    return IO.Input.read_int()

def input_string() -> str:
    return IO.Input.read_string()

def input_array() -> List[int]:
    return IO.Input.read_int_array()

def output(x) -> None:
    IO.output(x)

# Additional convenience functions for single word reading
def input_word() -> str:
    return _input_buffer.next_word()

def input_line() -> str:
    return _input_buffer.next_line()

def consume_newline() -> None:
    return IO.Input.consume_newline()

{{user template}}

{{Solution Class}}

def main():
    t = IO.Input.read_int()
    for _ in range(t):
{{Input Statements}}

if __name__ == "__main__":
    main()